xapian-applications/omega/handler_libebook.cc

```diff
@@ -0,0 +1,181 @@
+	const char* eol;
+	if (p)
+	    eol = p++;
+	else
+	    p = eol = end;
+	if ((end - start) > 5 && memcmp(start, "meta:", 5) == 0) {
+	    start += 5;
+	    switch (*start) {
+		case 'i': {
+		    if (author.empty())
+			PARSE_FIELD(start, eol, "initial-creator", author);
+		    break;
+		}
+		case 'k': {
+		    PARSE_FIELD(start, eol, "keyword", keywords);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 3 && memcmp(start, "dc:", 3) == 0) {
+	    start += 3;
+	    switch (*start) {
+		case 'c': {
+		    if (!author.empty())
+			author.clear();
+		    PARSE_FIELD(start, eol, "creator", author);
+		    break;
+		}
+		case 's': {
+		    PARSE_FIELD(start, eol, "subject", keywords);
+		    break;
+		}
+		case 't': {
+		    PARSE_FIELD(start, eol, "title", title);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 8 && memcmp(start, "dcterms:", 8) == 0) {
+	    start += 8;
+	    PARSE_FIELD(start, eol, "available", keywords);
+	}
+    }
+}
+
+static void
+clear_text(string& str, const char* text)
+{
+    for (int i = 0; text[i] != '\0'; ++i)
+	if (!isspace(text[i]) || (i && !isspace(text[i - 1])))
+	    str.push_back(text[i]);
+}
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages)
+{
+    try {
+	shared_ptr<RVNGInputStream> input;
+	RVNGString content_dump, metadata_dump;
+	const char* file = filename.c_str();
+	bool f_txt = false, f_meta = false;
+
+	if (RVNGDirectoryStream::isDirectory(file))
+	    input.reset(new RVNGDirectoryStream(file));
+	else
+	    input.reset(new RVNGFileStream(file));
+
+	EBOOKDocument::Type type = EBOOKDocument::TYPE_UNKNOWN;
+	EBOOKDocument::Confidence confidence =
+				EBOOKDocument::isSupported(input.get(), &type);
+
+	if (EBOOKDocument::CONFIDENCE_SUPPORTED_PART == confidence) {
+	    input.reset(RVNGDirectoryStream::createForParent(file));
+	    confidence = EBOOKDocument::isSupported(input.get(), &type);
+	}
+
+	if ((EBOOKDocument::CONFIDENCE_EXCELLENT != confidence) &&
+	    (EBOOKDocument::CONFIDENCE_WEAK != confidence)) {
+	    cerr << "Libe-book Error: The format is not supported" << endl;
+	    return false;
+	}
+
+	// Extract metadata if possible
+	RVNGTextTextGenerator metadata(metadata_dump, true);
+	if (EBOOKDocument::RESULT_OK ==
+	    EBOOKDocument::parse(input.get(), &metadata, type)) {
+	    const char* metadata = metadata_dump.cstr();
+	    size_t len = metadata_dump.size();
+	    parse_metadata(metadata, len, author, title, keywords);
+	    f_meta = true;
+	} else {
+	    cerr << "Libe-book Error: Fail to extract metadata" << endl;
+	}
+	(void)pages;
+	// Extract Dump if possible
+	RVNGTextTextGenerator content(content_dump, false);
+	if (EBOOKDocument::RESULT_OK ==
+	    EBOOKDocument::parse(input.get(), &content, type)) {
+	    clear_text(dump, content_dump.cstr());
+	    f_txt = true;
+	} else {
+	    cerr << "Libe-book Error: Fail to extract text" << endl;
+	}
+	return f_txt || f_meta;
```

Comment:
Is this actually the check you intended?  This will return "ok" if there's at most one error.

Looking at `enum Result` in `EBOOKDocument.h`, the results other than `RESULT_OK` look to be problems with accessing the file or with its contents (or encryption/password issues that we ought not hit because the confidence check above should have already caught those cases).  Those seem like problems that are better handled by giving up.

But if that is actually the check you wanted, why use two flag variables?  You'd only need a single `bool extracted_something = false;` and set that to `true` if anything works.

Reply:
I am going to change that.